### PR TITLE
ddl: add table option pre_split_regions for pre-split region when create table with  shard_row_id_bits.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Pre-split table region when creat table with  shard_row_id_bits.

If we do not pre-split region, there is only 1 region of the table at first, then maybe still have write hot spots at the beginning.

### What is changed and how it works?
1. Do pre-split region when create table with  `shard_row_id_bits` and `pre_split_regions`.  

2. scatter split region. Wait PD PR merge: https://github.com/pingcap/pd/pull/1501

Pre-split region will split 2^(pre_split_regions-1) for table row data, split 1 region for every index separately.

The pre-split process was executed asynchronous.

### Check List 

Tests 

 - Adding test...

Code changes

 - Has exported function/method change

Side effects

### CODE REVIEW

**Comment by winkyao** [comment]
Please update the desc `Adding test`, do you finish adding?

File: ddl/table.go [hunk 1]
```diff
@@ -335,6 +341,34 @@ func splitTableRegion(store kv.Storage, tableID int64) {
 	}
 }
 
+func preSplitTableRegion(store kv.Storage, tblInfo *model.TableInfo) {
+	s, ok := store.(splitableStore)
+	if !ok {
+		return
+	}
+	// Split table region.
+	step := int64(1 << (tblInfo.ShardRowIDBits - tblInfo.PreSplitRegions))
```
Comment (nolouch) [inline ddl/table.go]: The `PreSplitRegions` means the number of prefix bits used for splitting? the name confused me.

File: ddl/table.go [hunk 2]
```diff
@@ -335,6 +341,34 @@ func splitTableRegion(store kv.Storage, tableID int64) {
...
+	// The highest bit is the symbol bit， and alloc _tidb_rowid will always be positive number.
```
Comment (nolouch) [inline ddl/table.go]: s/，/,/

File: ddl/table.go [hunk 3]
```diff
@@ -335,6 +341,34 @@ func splitTableRegion(store kv.Storage, tableID int64) {
...
+	// So we only need to split the region for the positive number.
+	max := int64(1 << (tblInfo.ShardRowIDBits - 1))
+	for p := int64(step); p < max; p += step {
+		recordID := p << (64 - tblInfo.ShardRowIDBits)
```
Comment (nolouch) [inline ddl/table.go]: should consider the highest symbol bit?

**Comment by nolouch** [comment]
Can we try to detect the scattering operator is finished then return the DDL job is finished, like https://github.com/tikv/tikv/pull/4423/files#diff-14a21dd3ee41f0c7bb3547557e5c815cR223?

File: ddl/table.go (hunk 1 above)
Comment (crazycs520) [reply ddl/table.go]: PreSplitRegions means will pre-split 2^(PreSplitRegions - 1) regions.

File: ddl/table.go (hunk 3 above)
Comment (crazycs520) [reply ddl/table.go]: Already considered. 
```
supose:
ShardRowIDBits = 5
PreSplitRegions = 3

then will pre-split 2^(3-1) = 4 regions.

in this code: 
max = 1 << (tblInfo.ShardRowIDBits - 1) = 1 << (5-1) = 16
step := int64(1 << (tblInfo.ShardRowIDBits - tblInfo.PreSplitRegions)) = 1 << (5-3) = 4;

then split regionID is below:
4 << 59 = 2305843009213693952
8 << 59 = 4611686018427387904
12 << 59 = 6917529027641081856

the 4 regions range is below:
0                                        ~ 2305843009213693952
2305843009213693952 ~ 4611686018427387904
4611686018427387904 ~ 6917529027641081856
6917529027641081856 ~ 9223372036854775807 ( (1 << 63) - 1 )

and the max _tidb_rowid is 9223372036854775807, it won't be negative number.
```

Comment (nolouch) [reply ddl/table.go]: got it.

**Comment by bb7133** [comment]
Please update go.sum / go.mod for the parser

File: ddl/table.go [hunk 4]
```diff
@@ -74,6 +74,7 @@ func onCreateTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 			// TODO: Add restrictions to this operation.
 			go splitTableRegion(d.store, tbInfo.ID)
```
Comment (winkyao) [inline ddl/table.go]: Should we modify splitTableRegion strategy here?

File: ddl/table.go (hunk 1 above)
Comment (winkyao) [reply ddl/table.go]: The name is very confusing. maybe name it `PreSplitByShardbits`

File: ddl/table.go [hunk 5]
```diff
@@ -335,6 +338,50 @@ func splitTableRegion(store kv.Storage, tableID int64) {
...
+		return
+	}
+	regionIDs := make([]uint64, 0, 1<<(tblInfo.PreSplitRegions-1)+len(tblInfo.Indices))
+	// Split table region.
+	step := int64(1 << (tblInfo.ShardRowIDBits - tblInfo.PreSplitRegions))
+	// The highest bit is the symbol bit， and alloc _tidb_rowid will always be positive number.
+	// So we only need to split the region for the positive number.
+	max := int64(1 << (tblInfo.ShardRowIDBits - 1))
+	for p := int64(step); p < max; p += step {
+		recordID := p << (64 - tblInfo.ShardRowIDBits)
```
Comment (winkyao) [inline ddl/table.go]: Please add some examples as a comment here. like

```
supose:
ShardRowIDBits = 5
PreSplitRegions = 3

then will pre-split 2^(3-1) = 4 regions.

in this code: 
max = 1 << (tblInfo.ShardRowIDBits - 1) = 1 << (5-1) = 16
step := int64(1 << (tblInfo.ShardRowIDBits - tblInfo.PreSplitRegions)) = 1 << (5-3) = 4;

then split regionID is below:
4 << 59 = 2305843009213693952
8 << 59 = 4611686018427387904
12 << 59 = 6917529027641081856

the 4 regions range is below:
0                                        ~ 2305843009213693952
2305843009213693952 ~ 4611686018427387904
4611686018427387904 ~ 6917529027641081856
6917529027641081856 ~ 9223372036854775807 ( (1 << 63) - 1 )

and the max _tidb_rowid is 9223372036854775807, it won't be negative number.
```

File: store/tikv/backoff.go [hunk 6]
```diff
@@ -173,6 +173,8 @@ const (
 	deleteRangeOneRegionMaxBackoff = 100000
 	rawkvMaxBackoff                = 20000
 	splitRegionBackoff             = 20000
+	scatterRegionBackoff           = 20000
+	waitScatterRegionFinishBackoff = 60000
```
Comment (winkyao) [inline store/tikv/backoff.go]: scatter may need several minutes to finish, is 60s enough?

File: ddl/table.go (hunk 1 above)
Comment (crazycs520) [reply ddl/table.go]: ok.

File: ddl/table.go (hunk 4 above)
Comment (crazycs520) [reply ddl/table.go]: This PR is not to do this.

**Comment by nolouch** [comment]
Can we show the `pre-split-regions` in `show create table xxx`?

**Comment by crazycs520** [comment]
@nolouch Done.

File: store/tikv/split_region.go [hunk 7]
```diff
@@ -42,32 +48,100 @@ func (s *tikvStore) SplitRegion(splitKey kv.Key) error {
...
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+}
+
+func (s *tikvStore) SplitRegionAndScatter(splitKey kv.Key) (uint64, error) {
+	left, err := s.splitRegion(splitKey)
```
Comment (nolouch) [inline store/tikv/split_region.go]: If `splitRegion` failed will retry to split? such as `epoch is not matched`.

Comment (crazycs520) [reply store/tikv/split_region.go]: There is a `Backoff` in `splitRegion`, so it will retry.

**Pull request merged**